prow/test/integration/fakegitserver/fakegitserver.go

```diff
@@ -0,0 +1,545 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// fakegitserver serves Git repositories over HTTP for integration tests.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/cgi"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-git/go-billy/v5/osfs"
+	"github.com/go-git/go-billy/v5/util"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/cache"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/storage/filesystem"
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type options struct {
+	port                int
+	gitBinary           string
+	gitReposParentDir   string
+	populateSampleRepos bool
+	fooRepoRemoteURL    string
+}
+
+func (o *options) validate() error {
+	return nil
+}
+
+// flagOptions defines default options.
+func flagOptions() *options {
+	o := &options{}
+	flag.IntVar(&o.port, "port", 8888, "Port to listen on.")
+	flag.StringVar(&o.gitBinary, "git-binary", "/usr/bin/git", "Path to the `git` binary.")
+	flag.StringVar(&o.gitReposParentDir, "git-repos-parent-dir", "/git-repo", "Path to the parent folder containing all Git repos to serve over HTTP.")
+	flag.BoolVar(&o.populateSampleRepos, "populate-sample-repos", false, "Whether to populate /git-repo with hardcoded sample repos. Used for integration tests.")
+	flag.StringVar(&o.fooRepoRemoteURL, "foo-repo-remote-URL", "http://localhost:8888/repo/foo", "URL of foo repo, as a submodule from inside the bar repo. This is only used when -populate-sample-repos is given.")
+	return o
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := flagOptions()
+	flag.Parse()
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid arguments.")
+	}
+	defer interrupts.WaitForGracefulShutdown()
+
+	health := pjutil.NewHealth()
+	health.ServeReady()
+
+	r := mux.NewRouter()
+
+	// Only send requests under the /repo/... path to git-http-backend. This way
+	// we can have other paths (if necessary) to take in custom commands from
+	// integration tests (e.g., "/admin/reset" to reset all repos back to their
+	// original state).
+	r.PathPrefix("/repo").Handler(gitCGIHandler(o.gitBinary, o.gitReposParentDir))
+
+	server := &http.Server{
+		Addr:    fmt.Sprintf(":%d", o.port),
+		Handler: r,
+	}
+
+	if o.populateSampleRepos {
+		err := mkSampleRepos(o.gitReposParentDir, o.fooRepoRemoteURL)
```

Comment:
Don't think this should be blocking for them moment, but I think eventually we should not have the repo information hard coded in the fake git server. We will need to use the fake git server for multiple integration tests and it would be better if it was flexible enough to allow users to populate it after it has already started serving.

Reply:
> if it was flexible enough to allow users to populate it after it has already started serving

It is already flexible enough to do this. You can use volumes with pre-populated repos, or if you like you could just create a script and embed that into the K8s Deployment YAML (e.g., run some git commands and after that call FGS).

Reply:
> it will be easier to understand what is going on in the integration tests if the git server is populated there instead of by default.

The main reason I baked in these sample repos in here is because it made debugging the repos so much easier (where I could easily check how the repos were constructed without having to run the integration tests). It would take more code to move this logic out to the tests because the tests then have to worry about setup/teardown of these repos (i.e., the tests can't say "create repo foo in FGS" because it may have already been created before in a previous run). I don't think the additional engineering cost is worth the supposed readability gain.

Reply:
Agree that it's easier to debug for basic scenarios, so pre-populating a couple of seeds repos looks good to me.

Worth noting that for gerrit + inrepoconfig integration tests, the fake gerrit server and fake git server both play a role, and they will need to be consistent to each other. For example, for a given gerrit change:
1. fake gerrit server reports a gerrit change from: org-A/repo-B, changeNumber-100, changeID-def123 
2. prow/gerrit binary will contact fake git server to clone org-A/repo-B from HEAD, and merge latest SHA from changeNumber-100, changeID-def123.

In the above scenario, fake gerrit server and fake git server will both respond to the request for org-A/repo-B, changeNumber-100, changeID-def123 , and be consistent. Which so far can be done by pre-populated seed org-A/repo-B. For advanced scenarios, such as:
- Incremental test of new commit in main branch
- Incremental test of new commit in changeNumber-100
These scenarios might not be very well supported by pre-populated repo, and might need a way to sync and update fake gerrit server and fake git server.

Not meant to request change in this PR, pointing this out just for future reference. Since this PR lgtm:

/lgtm
/hold
If you want to wait for @cjwagner taking a look.

Reply:
Thinking a bit more about this, I have a feeling that adding in support to let users create repos on FGS isn't *that* much work. The simplest option is to let clients deal with the complexity --- that is, let the tests create the git repos they need (somewhere client-side) and then `git push` it up to FGS. The downside here as I mentioned before is that clients have to take care of housekeeping issues (making a temporary dir, creating the repo, making sure that git invocations are not polluted with the local machine's git environment settings, remembering to clean it up after the test, etc.).

As an alternative we can set up an endpoint `/setup-repo` on FGS that accepts a string (bash script) and just runs it server-side (the script would do all the work of `git init`, `git add`, `git commit`, etc). We're letting FGS become a target for a shell injection attack but I don't think that's a big concern because it's not a real production service. Anyway, doing everything server-side like this would reduce client-side complexity of having to deal with the housekeeping issues I mentioned up above.

I'm starting to lean toward this `/setup-repo` endpoint idea. Thoughts?

Reply:
So I just went ahead and implemented this already locally 😅 . I will port the test cases here to use it (and removed the mkSampleRepo* functions) --- this is somewhat mechanical and is straightforward to do. Stay tuned!

Reply:
OK so I've changed it so that the test case sends a shell script to populate the repo (each test case does this in its test case configuration). I've also added more verbosity when the test cases encounter a failure. PTAL @mpherman2